src/coreclr/tools/SuperFileCheck/Program.cs

```diff
@@ -0,0 +1,623 @@
+                        else
+                        {
+                            stdOut.AppendLine(e.Data);
+                        }
+                    };
+
+                    proc.ErrorDataReceived += (_, e) =>
+                    {
+                        if (e.Data == null)
+                        {
+                            errorWaitHandle.Set();
+                        }
+                        else
+                        {
+                            stdErr.AppendLine(e.Data);
+                        }
+                    };
+
+                    proc.BeginOutputReadLine();
+                    proc.BeginErrorReadLine();
+
+                    await proc.WaitForExitAsync();
+                    outputWaitHandle.WaitOne();
+                    errorWaitHandle.WaitOne();
+
+                    var exitCode = proc.ExitCode;
+                    return new FileCheckResult(exitCode, stdOut.ToString(), stdErr.ToString());
+                }
+            }
+            catch (Exception ex)
+            {
+                return new FileCheckResult(1, String.Empty, ex.Message);
+            }
+        }
+
+        /// <summary>
+        /// Get the method name from the method declaration.
+        /// </summary>
+        static string GetMethodName(MethodDeclarationSyntax methodDecl)
+        {
+            return
+                methodDecl.ChildTokens()
+                .OfType<SyntaxToken>()
+                .Where(x => x.IsKind(SyntaxKind.IdentifierToken)).First().ValueText;
+        }
+
+        static IEnumerable<SyntaxTrivia> GetDescendantSingleLineCommentTrivia(SyntaxNode node)
+        {
+            return
+                node
+                .DescendantTrivia()
+                .Where(x => x.IsKind(SyntaxKind.SingleLineCommentTrivia));
+        }
+
+        /// <summary>
+        /// Gather all syntactical method declarations whose body contains
+        /// FileCheck syntax.
+        /// </summary>
+        static MethodDeclarationInfo[] FindMethodsByFile(string filePath, string[] checkPrefixes)
+        {
+            var syntaxTree = CSharpSyntaxTree.ParseText(SourceText.From(File.ReadAllText(filePath)));
+            var root = syntaxTree.GetRoot();
+
+            var trivia =
+                GetDescendantSingleLineCommentTrivia(root)
+                .Where(x =>
+                {
+                    if (x.Token.Parent != null)
+                    {
+                        return !x.Token.Parent.Ancestors().Any(p => p.IsKind(SyntaxKind.MethodDeclaration) && p.Span.Contains(x.Span));
+                    }
+                    else
+                    {
+                        return true;
+                    }
+                })
+                .Where(x => ContainsCheckPrefixes(x.ToString(), checkPrefixes))
+                .ToArray();
+
+            if (trivia.Length > 0)
+            {
+                throw new Exception("FileCheck syntax not allowed outside of a method.");
```

Comment:
We should probably define an exception instead of using `Exception`.

Reply:
Sure, I'll create a `SuperFileCheckException`.